internal/rpc/sasl_dialer.go

```diff
@@ -0,0 +1,43 @@
+package rpc
+
+import (
+	"context"
+	"net"
+
+	"github.com/colinmarc/hdfs/v2/internal/protocol/hadoop_common"
+)
+
+// DatanodeSaslDialer is a dialer that will use the Privacy and Integrity
+// data members in order to decide whether or not it needs to wrap the connection
+// to the datanode for proper handling of encryption / integrity protection
+// after it's been constructed it can be used as one would use net.Dialer
+type DatanodeSaslDialer struct {
+	Dialer    func(ctx context.Context, network, addr string) (net.Conn, error)
+	KeyFunc   func() *EncryptionKey
```

Comment:
This is a bit dicey. I'd prefer we call this function once and then reuse the result.

Reply:
fair, I can call it in `Client` when we create this and store the result rather than passing the function itself.

Reply:
hm, maybe just pass the two options into `NamenodeConnection` and then put the `wrapDatanodeDialer` func there?

Reply:
whoops, meant that in response to the above comment